.github/workflows/rust-prs.yml

```diff
@@ -0,0 +1,459 @@
+          cat $GITHUB_OUTPUT
+
+      - name: Check paths changed
+        uses: dorny/paths-filter@v3
+        id: paths-changed
+        with:
+          filters: |
+            grafbase-docker:
+              - cli/Dockerfile
+            gateway-docker:
+              - gateway/Dockerfile
+
+  check-format:
+    needs: [what-changed]
+    if: fromJson(needs.what-changed.outputs.rust).changed-packages
+    runs-on: ubicloud-standard-2
+    steps:
+      - name: Get sources
+        uses: actions/checkout@v4
+
+      - name: Extract the Rust version to use from the `rust-toolchain.toml` file
+        shell: bash
+        run: |
+          rust_version=$(grep "channel" rust-toolchain.toml | cut -d "=" -f 2 | cut -d " " -f 2)
+          echo "RUST_VERSION=$rust_version" >> $GITHUB_ENV
+
+      - name: Install Rust
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ env.RUST_VERSION }}
+          components: rustfmt
+
+      - name: cargo fmt
+        shell: bash
+        run: |
+          cargo fmt --check
+
+  example-component:
+    if: contains(fromJson(needs.what-changed.outputs.rust).changed-packages, 'integration-tests')
+    needs: [what-changed]
+    runs-on: ubicloud-standard-8
+    steps:
+      - name: Get sources
+        uses: actions/checkout@v4
+
+      - name: Install cargo binstall
+        uses: cargo-bins/cargo-binstall@v1.10.3
+
+      - name: Install cargo-component
+        shell: bash
+        run: |
+          cargo binstall cargo-component --secure -y
+
+      - name: Build the WASI components for tests
+        shell: bash
+        working-directory: engine/crates/wasi-component-loader/examples
+        run: cargo component build
+
+      - uses: actions/upload-artifact@v4
+        with:
+          name: example-component
+          path: engine/crates/wasi-component-loader/examples/target/wasm32-wasip1/debug/*.wasm
+          retention-days: 5
+
+  cli-assets:
+    if: contains(fromJson(needs.what-changed.outputs.rust).changed-packages, 'grafbase')
+    needs: [what-changed]
+    runs-on: ubicloud-standard-8
+    steps:
+      - name: Get sources
+        uses: actions/checkout@v4
+
+      - name: Build assets
+        uses: ./.github/actions/cli_assets
+
+      - name: Copy ~/.grafbase for upload
+        shell: bash
+        run: |
+          mkdir home-dot-grafbase
+          cp -r ~/.grafbase/* home-dot-grafbase/
+
+      - uses: actions/upload-artifact@v4
+        with:
+          name: home-dot-grafbase
+          path: home-dot-grafbase
+          retention-days: 5
+
+      - uses: actions/upload-artifact@v4
+        with:
+          name: cli-app-dist
+          path: packages/cli-app/dist
+          retention-days: 5
+
+      - uses: actions/upload-artifact@v4
+        with:
+          name: wrappers-dist.js
+          path: cli/wrappers/dist.js
+          retention-days: 5
+
+  builds:
+    needs: [what-changed, example-component, cli-assets]
+    if: |
+      (!cancelled())
```

Comment:
Is there a clear reason why this is only needed for this job?

Reply:
`example-component` & `cli-assets` are both optional - without doing some `if` shenanigans this wouldn't run if they were skipped.  

I did have `always()` initially, but it turns out if you put `always()` in a job it becomes uncancellable - this is the recommended alterantive that lets you cancel.

I'll add a comment though